demos/python_demos/common/images_capture.py

```diff
@@ -0,0 +1,155 @@
+        self.file_id = 0
+        for name in self.names:
+            filename = os.path.join(self.dir, name)
+            image = cv2.imread(filename, cv2.IMREAD_COLOR)
+            if image is not None:
+                return
+        raise OpenError("Can't read the first image from {}".format(input))
+
+    def read(self):
+        while self.file_id < len(self.names):
+            filename = os.path.join(self.dir, self.names[self.file_id])
+            image = cv2.imread(filename, cv2.IMREAD_COLOR)
+            self.file_id += 1
+            if image is not None:
+                return image
+        if self.loop:
+            self.file_id = 0
+            while self.file_id < len(self.names):
+                filename = os.path.join(self.dir, self.names[self.file_id])
+                image = cv2.imread(filename, cv2.IMREAD_COLOR)
+                self.file_id += 1
+                if image is not None:
+                    return image
+        return None
+
+    def fps(self):
+        return 1.0
+
+    def get_type(self):
+        return 'DIR'
+
+
+class VideoCapWrapper(ImagesCapture):
+
+    def __init__(self, input, loop, camera_resolution):
+        super().__init__(loop)
+        self.cap = cv2.VideoCapture()
+        try:
+            status = self.cap.open(int(input))
+            self.type = 'CAMERA'
+            self.cap.set(cv2.CAP_PROP_BUFFERSIZE, 1)
+            self.cap.set(cv2.CAP_PROP_FRAME_WIDTH, camera_resolution[0])
+            self.cap.set(cv2.CAP_PROP_FRAME_HEIGHT, camera_resolution[1])
+            self.cap.set(cv2.CAP_PROP_AUTOFOCUS, 1)
+            self.cap.set(cv2.CAP_PROP_FOURCC, cv2.VideoWriter_fourcc(*'MJPG'))
+        except ValueError:
+            status = self.cap.open(input)
+            self.type = 'VIDEO'
+        if not status:
+            raise InvalidInput("Can't open the {} from {}".format(self.get_type().lower(), input))
+
+    def read(self):
+        status, image = self.cap.read()
+        if not status:
+            if not self.loop:
+                return None
+            self.cap.set(cv2.CAP_PROP_POS_FRAMES, 0)
+            status, image = self.cap.read()
+            if not status:
+                return None
+        return image
+
+    def fps(self):
+        return self.cap.get(cv2.CAP_PROP_FPS)
+
+    def get_type(self):
+        return self.type
+
+
+def open_images_capture(input, loop, camera_resolution=(1280, 720)):
+    errors = {InvalidInput: [], OpenError: []}
+    for reader in (ImreadWrapper, DirReader):
+        try:
+            return reader(input, loop)
+        except (InvalidInput, OpenError) as e:
+            errors[type(e)].append(e.message)
+    try:
+        return VideoCapWrapper(input, loop, camera_resolution)
+    except (InvalidInput, OpenError) as e:
+        errors[type(e)].append(e.message)
+    if not errors[OpenError]:
+        print(*errors[InvalidInput], file=sys.stderr, sep='\n')
+    else:
+        print(*errors[OpenError], file=sys.stderr, sep='\n')
```

Comment:
This all seems really involved. Why not follow the C++ implementation? Try each implementation ignoring `InvalidError`, and if all of them fail, throw an exception.

Reply:
Firstly, the python version had followed the c++ implementation, but I changed it according to Eduard request about errors https://github.com/openvinotoolkit/open_model_zoo/pull/1815#discussion_r528633557.